File: jvm/core/main/scala/com/webank/eggroll/core/clustermanager/metadata/StoreCrudOperator.scala
/*
 * Copyright (c) 2019 - now, Eggroll Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *
 */

package com.webank.eggroll.core.clustermanager.metadata

import java.{lang, util}
import java.util.concurrent.ConcurrentHashMap

import com.webank.eggroll.core.clustermanager.constant.RdbConstants
import com.webank.eggroll.core.constant.{BindingStrategies, PartitionStatus, ServerNodeStatus, ServerNodeTypes, SessionConfKeys, StoreStatus, StringConstants}
import com.webank.eggroll.core.error.CrudException
import com.webank.eggroll.core.meta._
import com.webank.eggroll.core.util.Logging
import com.webank.eggroll.core.clustermanager.dao.generated.mapper.{ServerNodeMapper, StoreLocatorMapper, StorePartitionMapper}
import com.webank.eggroll.core.clustermanager.dao.generated.model._
import com.webank.eggroll.core.clustermanager.session.SessionManager
import org.apache.ibatis.session.SqlSession

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

class StoreCrudOperator extends CrudOperator with Logging {
  private val crudOperatorTemplate = new CrudOperatorTemplate()
  def getOrCreateStore(input: ErStore): ErStore = {
    def doGetOrCreateStore(input: ErStore, sqlSession: SqlSession): ErStore = {
      val existing = StoreCrudOperator.doGetStore(input, sqlSession)
      if (existing != null) {
        existing
      } else {
        StoreCrudOperator.doCreateStore(input, sqlSession)
      }
    }

    crudOperatorTemplate.doCrudOperationSingleResult(doGetOrCreateStore, input, openTransaction = true)
  }

  def getStore(input: ErStore): ErStore = {
    crudOperatorTemplate.doCrudOperationSingleResult(StoreCrudOperator.doGetStore, input)
  }

  def deleteStore(input: ErStore): ErStore = {
    crudOperatorTemplate.doCrudOperationSingleResult(StoreCrudOperator.doDeleteStore, input, openTransaction = true)
  }
}

object StoreCrudOperator {
  private val nodeIdToNode = new ConcurrentHashMap[java.lang.Long, ServerNode]()
  private[metadata] def doGetStore(input: ErStore, sqlSession: SqlSession): ErStore = {
    val inputOptions = input.options
    val sessionId = inputOptions.getOrDefault(SessionConfKeys.CONFKEY_SESSION_ID, StringConstants.UNKNOWN)

    // getting input locator
    val inputStoreLocator = input.storeLocator
    val storeLocatorExample = new StoreLocatorExample
    storeLocatorExample.createCriteria()
      .andStoreTypeEqualTo(inputStoreLocator.storeType)
      .andNamespaceEqualTo(inputStoreLocator.namespace)
      .andNameEqualTo(inputStoreLocator.name)
      .andStatusEqualTo(StoreStatus.NORMAL)
    val storeLocatorMapper = sqlSession.getMapper(classOf[StoreLocatorMapper])

    val storeLocatorResult = storeLocatorMapper.selectByExampleWithRowbounds(storeLocatorExample, RdbConstants.SINGLE_ROWBOUND)

    if (storeLocatorResult.isEmpty) {
      return null
    }

    val store = storeLocatorResult.get(0)
    val storeLocatorId = store.getStoreLocatorId

    // getting partitions
    val storePartitionExample = new StorePartitionExample
    storePartitionExample.createCriteria()
      .andStoreLocatorIdEqualTo(storeLocatorId)

    storeLocatorExample.setOrderByClause("store_partition_id asc")

    val storePartitionMapper = sqlSession.getMapper(classOf[StorePartitionMapper])
    val storePartitionResult = storePartitionMapper.selectByExample(storePartitionExample)

    if (storePartitionResult.isEmpty) {
      throw new IllegalStateException("store locator found but no partition found")
    }

    val missingNodeId = new util.HashSet[java.lang.Long](storePartitionResult.size())
    val partitionAtNodeIds = ArrayBuffer[Long]()

    storePartitionResult.forEach(r => {
      val nodeId = r.getNodeId
      if (!nodeIdToNode.containsKey(nodeId)) missingNodeId.add(nodeId)
      partitionAtNodeIds += nodeId
    })

    if (!missingNodeId.isEmpty) {
      val nodeExample = new ServerNodeExample
      nodeExample.createCriteria()
        .andServerNodeIdIn(new util.ArrayList[lang.Long](missingNodeId))
        .andNodeTypeEqualTo(ServerNodeTypes.NODE_MANAGER)
        .andStatusEqualTo(ServerNodeStatus.HEALTHY)

      val nodeMapper = sqlSession.getMapper(classOf[ServerNodeMapper])
      val nodeResult = nodeMapper.selectByExample(nodeExample)

      if (nodeResult.isEmpty) {
        throw new IllegalStateException(s"No valid node for this store: ${inputStoreLocator}")
      }

      nodeResult.forEach(n => {nodeIdToNode.putIfAbsent(n.getServerNodeId, n)})
    }

    val outputStoreLocator = ErStoreLocator(
      storeType = store.getStoreType,
      namespace = store.getNamespace,
      name = store.getName,
      path = store.getPath,
      totalPartitions = store.getTotalPartitions,
      partitioner = store.getPartitioner,
      serdes = store.getSerdes
    )

    val outputOptions = new ConcurrentHashMap[String, String]()
    if (inputOptions != null) {
      outputOptions.putAll(inputOptions)
    }

    val existingBinding = SessionManager.getBindingPlan(
      sessionId = sessionId,
      totalPartitions = outputStoreLocator.totalPartitions,
      serverNodeIds = partitionAtNodeIds.toArray)

    // process output partitions
    val outputPartitions = ArrayBuffer[ErPartition]()
    if (existingBinding != null
      && existingBinding.id.equals(inputOptions.get(SessionConfKeys.CONFKEY_SESSION_EGG_BINDING_ID))
      && !inputOptions.getOrDefault(SessionConfKeys.CONFKEY_SESSION_EGG_BINDING_INCLUDE_DETAILS, "false").toBoolean) {
      outputOptions.put(SessionConfKeys.CONFKEY_SESSION_EGG_BINDING_ID, existingBinding.id)
    } else {
      val bindingId = ErPartitionBindingPlan.genId(
        sessionId = sessionId,
        totalPartitions = outputStoreLocator.totalPartitions,
        serverNodeIds = partitionAtNodeIds.toArray,
        strategy = inputOptions.getOrDefault(SessionConfKeys.CONFKEY_SESSION_EGG_BINDING_STRATEGY, BindingStrategies.ROUND_ROBIN))
      outputOptions.put(SessionConfKeys.CONFKEY_SESSION_EGG_BINDING_ID, bindingId)
      outputOptions.put(SessionConfKeys.CONFKEY_SESSION_EGG_BINDING_STRATEGY, BindingStrategies.ROUND_ROBIN)

      val newBinding = SessionManager.createBindingPlan(
        sessionId = sessionId,
        totalPartitions = outputStoreLocator.totalPartitions,
        serverNodeIds = partitionAtNodeIds.toArray)
      outputPartitions ++= newBinding.toPartitions()
    }

    ErStore(storeLocator = outputStoreLocator, partitions = outputPartitions.toArray, options = outputOptions)
  }

  private[metadata] def doCreateStore(input: ErStore, sqlSession: SqlSession): ErStore = {
    val inputOptions = input.options
    val sessionId = inputOptions.getOrDefault(SessionConfKeys.CONFKEY_SESSION_ID, StringConstants.UNKNOWN)

    val inputStoreLocator = input.storeLocator
    val newStoreLocator = new StoreLocator

    newStoreLocator.setStoreType(inputStoreLocator.storeType)
    newStoreLocator.setNamespace(inputStoreLocator.namespace)
    newStoreLocator.setName(inputStoreLocator.name)
    newStoreLocator.setPath(inputStoreLocator.path)
    newStoreLocator.setTotalPartitions(inputStoreLocator.totalPartitions)
    newStoreLocator.setPartitioner(inputStoreLocator.partitioner)
    newStoreLocator.setSerdes(inputStoreLocator.serdes)
    newStoreLocator.setStatus(StoreStatus.NORMAL)

    val storeLocatorMapper = sqlSession.getMapper(classOf[StoreLocatorMapper])
    var rowsAffected = storeLocatorMapper.insertSelective(newStoreLocator)
    if (rowsAffected != 1) {
      throw new CrudException(s"Illegal rows affected returned when creating store locator: ${rowsAffected}")
    }

    val partitionServerNodeBindingId = inputOptions.getOrDefault(SessionConfKeys.CONFKEY_SESSION_EGG_BINDING_ID, null)
    var outputTotalPartitions = inputStoreLocator.totalPartitions

    val finalPartitions: ArrayBuffer[ErPartition] = ArrayBuffer[ErPartition]()
    finalPartitions.sizeHint(inputStoreLocator.totalPartitions)

    // todo: find existing binding with same partition numbers
    val serverNodes: Array[ErServerNode] = if (partitionServerNodeBindingId != null) {
      val binding = SessionManager.getBindingPlan(sessionId = sessionId, bindingId = partitionServerNodeBindingId)
      val existingBinding = ArrayBuffer[ErServerNode]()
      existingBinding.sizeHint(binding.totalPartitions)
      binding.partitionToServerNodes.foreach(nid => {
        val node = nodeIdToNode.get(nid)

        // todo: conversion move to util or ErServerNode
        existingBinding += ErServerNode(
          id = node.getServerNodeId,
          name = node.getName,
          clusterId = node.getServerClusterId,
          endpoint = ErEndpoint(host = node.getHost, port = node.getPort),
          nodeType = node.getNodeType,
          status = node.getStatus)

      })
      outputTotalPartitions = binding.totalPartitions
      existingBinding.toArray
    } else if (SessionManager.getSessionDeployment(sessionId) != null && !sessionId.equals(StringConstants.UNKNOWN)) {
      SessionManager.getSessionDeployment(sessionId).serverCluster.serverNodes
    } else {
      ServerNodeCrudOperator.doGetServerNodes(input = ErServerNode(nodeType = ServerNodeTypes.NODE_MANAGER, status = ServerNodeStatus.HEALTHY),
        sqlSession = sqlSession)
    }

    val nodesCount = serverNodes.length
    val specifiedPartitions = input.partitions
    val isPartitionsSpecified = specifiedPartitions.length > 0
    if (isPartitionsSpecified) {
      throw new UnsupportedOperationException("specify partition not supported yet")
    }

    if (outputTotalPartitions <= 0) outputTotalPartitions = nodesCount << 2
    val storePartitionMapper = sqlSession.getMapper(classOf[StorePartitionMapper])

    val serverNodeIds = ArrayBuffer[Long]()
    for (i <- 0 until outputTotalPartitions) {
      val storePartitionRecord = new StorePartition
      val node: ErServerNode = serverNodes(i % nodesCount)

      storePartitionRecord.setStoreLocatorId(newStoreLocator.getStoreLocatorId)
      storePartitionRecord.setNodeId(node.id)
      storePartitionRecord.setPartitionId(i)
      storePartitionRecord.setStatus(PartitionStatus.PRIMARY)

      rowsAffected = storePartitionMapper.insertSelective(storePartitionRecord)
      if (rowsAffected != 1) {
        throw new CrudException(s"Illegal rows affected returned when creating store partition: ${rowsAffected}")
      }

      serverNodeIds += node.id
      // todo: bind with active session
      finalPartitions += ErPartition(
        id = i,
        storeLocator = inputStoreLocator,
        processor = ErProcessor(id = i,
          serverNodeId = storePartitionRecord.getNodeId,
          tag = "binding"))
    }

    val finalOptions = new ConcurrentHashMap[String, String]()
    if (inputOptions != null) finalOptions.putAll(inputOptions)
    val result = if (partitionServerNodeBindingId == null
      || inputOptions.getOrDefault(SessionConfKeys.CONFKEY_SESSION_EGG_BINDING_INCLUDE_DETAILS, "false").toBoolean) {
      val partitionToNodeIds = finalPartitions.map(p => p.processor.serverNodeId).toArray

      val binding = SessionManager.createBindingPlan(
        sessionId = sessionId,
        totalPartitions = outputTotalPartitions,
        serverNodeIds = partitionToNodeIds,
        strategy = BindingStrategies.ROUND_ROBIN)


      finalOptions.put(SessionConfKeys.CONFKEY_SESSION_EGG_BINDING_ID, binding.id)
      finalOptions.put(SessionConfKeys.CONFKEY_SESSION_EGG_BINDING_STRATEGY, BindingStrategies.ROUND_ROBIN)
      ErStore(
        storeLocator = inputStoreLocator,
        partitions = finalPartitions.toArray,
        options = finalOptions)
    } else {
      ErStore(
        storeLocator = inputStoreLocator,
        options = finalOptions)
    }

    result
  }

  private[metadata] def doDeleteStore(input: ErStore, sqlSession: SqlSession): ErStore = {
    val inputStoreLocator = input.storeLocator

    val storeLocatorExample = new StoreLocatorExample
    storeLocatorExample.createCriteria()
      .andStoreTypeEqualTo(inputStoreLocator.storeType)
      .andNamespaceEqualTo(inputStoreLocator.namespace)
      .andNameEqualTo(inputStoreLocator.name)
      .andStatusEqualTo(StoreStatus.NORMAL)
    val storeMapper = sqlSession.getMapper(classOf[StoreLocatorMapper])

    val storeResult = storeMapper.selectByExampleWithRowbounds(storeLocatorExample, RdbConstants.SINGLE_ROWBOUND)

    if (storeResult.isEmpty) {
      return null
    }

    val now = System.currentTimeMillis()
    val store = storeResult.get(0)
    store.setName(s"${store.getName}.${now}")
    store.setStatus(StoreStatus.DELETED)

    val rowsAffected = storeMapper.updateByPrimaryKeySelective(store)

    if (rowsAffected != 1) {
      throw new CrudException(s"Illegal rows affected returned when deleting store: ${rowsAffected}")
    }

    val outputStoreLocator = inputStoreLocator.copy(name = store.getName)

    ErStore(storeLocator = outputStoreLocator)
  }
}

File: jvm/cluster_manager/src/main/java/org/fedai/eggroll/clustermanager/dao/impl/StoreCrudOperator.java
package org.fedai.eggroll.clustermanager.dao.impl;

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.baomidou.mybatisplus.core.conditions.update.UpdateWrapper;
import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import com.google.inject.Inject;
import com.google.inject.Singleton;
import org.apache.commons.lang3.ObjectUtils;
import org.apache.commons.lang3.StringUtils;
import org.fedai.eggroll.clustermanager.entity.ServerNode;
import org.fedai.eggroll.clustermanager.entity.StoreLocator;
import org.fedai.eggroll.clustermanager.entity.StoreOption;
import org.fedai.eggroll.clustermanager.entity.StorePartition;
import org.fedai.eggroll.core.config.Dict;
import org.fedai.eggroll.core.constant.StringConstants;
import org.fedai.eggroll.core.context.Context;
import org.fedai.eggroll.core.exceptions.CrudException;
import org.fedai.eggroll.core.pojo.*;
import org.fedai.eggroll.core.utils.JsonUtil;
import org.fedai.eggroll.core.utils.LockUtils;
import org.mybatis.guice.transactional.Transactional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.ReentrantLock;
import java.util.stream.Collectors;

@Singleton
public class StoreCrudOperator {

    Logger logger = LoggerFactory.getLogger(StoreCrudOperator.class);
    @Inject
    StoreLocatorService storeLocatorService;
    @Inject
    StorePartitionService storePartitionService;
    @Inject
    ServerNodeService serverNodeService;
    @Inject
    StoreOptionService storeOptionService;

    private final Map<Long, Object> nodeIdToNode = new ConcurrentHashMap<>();
    Cache<String, ReentrantLock> storeLockCache = CacheBuilder.newBuilder()
            .maximumSize(1024)
            .expireAfterAccess(Long.MAX_VALUE, TimeUnit.NANOSECONDS)
            .build();


    public ErStore doGetStore(ErStore input) {
        Map<String, String> inputOptions = input.getOptions();

        // getting input locator
        ErStoreLocator inputStoreLocator = input.getStoreLocator();

        StoreLocator params = new StoreLocator();
        params.setNamespace(inputStoreLocator.getNamespace());
        params.setName(inputStoreLocator.getName());
        params.setStatus(Dict.NORMAL);

        if (!StringUtils.isBlank(inputStoreLocator.getStoreType())) {
            params.setStoreType(inputStoreLocator.getStoreType());
        }
        List<StoreLocator> storeLocatorResult = storeLocatorService.list(new QueryWrapper<>(params));
        if (storeLocatorResult.isEmpty()) {
            return null;
        }
        StoreLocator store = storeLocatorResult.get(0);
        Long storeLocatorId = store.getStoreLocatorId();

        QueryWrapper<StorePartition> queryStorePartition = new QueryWrapper<>();
        queryStorePartition.lambda().eq(StorePartition::getStoreLocatorId, storeLocatorId)
                .orderByAsc(StorePartition::getStorePartitionId);
        List<StorePartition> storePartitionResult = storePartitionService.list(queryStorePartition);
        if (storePartitionResult.isEmpty()) {
            throw new IllegalStateException("store locator found but no partition found");
        }

        List<Long> missingNodeId = new ArrayList<>();
        List<Long> partitionAtNodeIds = new ArrayList<>();

        for (StorePartition storePartition : storePartitionResult) {
            Long nodeId = storePartition.getNodeId();
            if (!nodeIdToNode.containsKey(nodeId)) {
                missingNodeId.add(nodeId);
            }
            partitionAtNodeIds.add(nodeId);
        }
        if (!missingNodeId.isEmpty()) {
            QueryWrapper<ServerNode> queryServerNode = new QueryWrapper<>();
            queryServerNode.lambda().eq(ServerNode::getStatus, Dict.HEALTHY)
                    .eq(ServerNode::getNodeType, Dict.NODE_MANAGER)
                    .in(ServerNode::getServerNodeId, missingNodeId);
            List<ServerNode> nodeResult = serverNodeService.list(queryServerNode);
            if (nodeResult.isEmpty()) {
                logger.error("No valid node for this store: " + JsonUtil.object2Json(inputStoreLocator) + " missingId " + missingNodeId.toString());
                throw new IllegalStateException("No valid node for this store: " + JsonUtil.object2Json(inputStoreLocator));
            }

            for (ServerNode serverNode : nodeResult) {
                Long serverNodeId = serverNode.getServerNodeId();
                nodeIdToNode.putIfAbsent(serverNodeId, serverNode);
            }
        }

        ErStoreLocator outputStoreLocator = new ErStoreLocator(store.getStoreLocatorId()
                , store.getStoreType()
                , store.getNamespace()
                , store.getName()
                , store.getPath()
                , store.getTotalPartitions()
                , store.getPartitioner()
                , store.getSerdes());
        QueryWrapper<StoreOption> storeOptionWrapper = new QueryWrapper<>();
        storeOptionWrapper.lambda().eq(StoreOption::getStoreLocatorId, storeLocatorId);
        List<StoreOption> storeOpts = storeOptionService.list(storeOptionWrapper);
        ConcurrentHashMap<String, String> outputOptions = new ConcurrentHashMap<>();
        if (inputOptions != null) {
            outputOptions.putAll(inputOptions);
        }
        if (storeOpts != null) {
            for (StoreOption r : storeOpts) {
                outputOptions.put(r.getName(), r.getData());
            }
        }

        List<ErPartition> outputPartitions = new ArrayList<>();
        for (StorePartition p : storePartitionResult) {

            ErProcessor erProcessor = new ErProcessor();
            erProcessor.setId(p.getPartitionId().longValue());
            erProcessor.setServerNodeId(p.getNodeId());

            outputPartitions.add(new ErPartition(p.getPartitionId()
                    , outputStoreLocator
                    , erProcessor
                    , -1
            ));
        }
        return new ErStore(outputStoreLocator, outputPartitions, outputOptions);
    }

    public ErStore doGetOrCreateStore(Context context, ErStore input) {
        ErStoreLocator inputStoreLocator = input.getStoreLocator();
        context.putLogData("name",inputStoreLocator.getName());
        context.putLogData("namespace",inputStoreLocator.getNamespace());
        String inputStoreType = inputStoreLocator.getStoreType();
        ErStore inputWithoutType = ObjectUtils.clone(input);
        inputWithoutType.getStoreLocator().setStoreType(StringConstants.EMPTY);
        try {
            LockUtils.lock(storeLockCache, input.getStoreLocator().buildKey());
//            logger.info("==============> store getorcreate lock key ={}",input.getStoreLocator().buildKey());
            ErStore existing = doGetStore(inputWithoutType);
            if (existing != null) {
                if (!existing.getStoreLocator().getStoreType().equals(inputStoreType)) {
                    logger.warn("store namespace: " + inputStoreLocator.getNamespace() + ", name: " +
                            inputStoreLocator.getName() + " already exist with store type: " +
                            existing.getStoreLocator().getStoreType() + ". requires type: " +
                            inputStoreLocator.getStoreType());
                }
                return existing;
            } else {
                return doCreateStore(input);
            }
        } finally {
            LockUtils.unLock(storeLockCache, input.getStoreLocator().buildKey());
//            logger.info("==============> store getorcreate uuuuuuuuuuuunlock key ={}",input.getStoreLocator().buildKey());
        }
    }

    public ErStore doCreateStore(ErStore input) {
        Map<String, String> inputOptions = input.getOptions();
        ErStoreLocator inputStoreLocator = input.getStoreLocator();

        StoreLocator newStoreLocator = new StoreLocator();
        newStoreLocator.setStoreType(inputStoreLocator.getStoreType());
        newStoreLocator.setNamespace(inputStoreLocator.getNamespace());
        newStoreLocator.setName(inputStoreLocator.getName());
        newStoreLocator.setPath(inputStoreLocator.getPath());
        newStoreLocator.setTotalPartitions(inputStoreLocator.getTotalPartitions());
        newStoreLocator.setPartitioner(inputStoreLocator.getPartitioner());
        newStoreLocator.setSerdes(inputStoreLocator.getSerdes());
        newStoreLocator.setStatus(Dict.NORMAL);
        boolean addStoreLocatorFlag = storeLocatorService.save(newStoreLocator);
        if (!addStoreLocatorFlag) {
            throw new CrudException("Illegal rows affected returned when creating store locator: 0");
        }

        int newTotalPartitions = inputStoreLocator.getTotalPartitions();
        List<ErPartition> newPartitions = new ArrayList<>();
        List<ErServerNode> serverNodes = serverNodeService.getListByErServerNode(
                new ErServerNode(
                        Dict.NODE_MANAGER
                        , Dict.HEALTHY)).stream().sorted(Comparator.comparingLong(ErServerNode::getId)).collect(Collectors.toList());
        int nodesCount = serverNodes.size();
        List<ErPartition> specifiedPartitions = input.getPartitions();
        boolean isPartitionsSpecified = specifiedPartitions.size() > 0;
        if (newTotalPartitions <= 0) {
            newTotalPartitions = nodesCount << 2;
        }
        if(nodesCount == 0){
            throw new RuntimeException("no health server node");
        }
        ArrayList<Long> serverNodeIds = new ArrayList<>();
        for (int i = 0; i < newTotalPartitions; i++) {
            ErServerNode node = serverNodes.get(i % nodesCount);
            StorePartition nodeRecord = new StorePartition();
            nodeRecord.setStoreLocatorId(newStoreLocator.getStoreLocatorId());
            nodeRecord.setNodeId(isPartitionsSpecified ? input.getPartitions().get(i % specifiedPartitions.size()).getProcessor().getServerNodeId() : node.getId());
            nodeRecord.setPartitionId(i);
            nodeRecord.setStatus(Dict.PRIMARY);
            boolean addNodeRecordFlag = storePartitionService.save(nodeRecord);
            if (!addNodeRecordFlag) {

                throw new CrudException("Illegal rows affected when creating node: 0");
            }
            serverNodeIds.add(node.getId());

            ErProcessor binding = new ErProcessor();
            binding.setId((long) i);
            binding.setServerNodeId(isPartitionsSpecified ? input.getPartitions().get(i % specifiedPartitions.size()).getProcessor().getServerNodeId() : node.getId());
            binding.setTag("binding");

            newPartitions.add(new ErPartition(i
                    , inputStoreLocator
                    , binding
                    , -1));
        }
        ConcurrentHashMap<String, String> newOptions = new ConcurrentHashMap<>();
        if (inputOptions != null) {
            newOptions.putAll(inputOptions);
        }
        for (Map.Entry<String, String> entry : newOptions.entrySet()) {
            StoreOption newStoreOption = new StoreOption();
            newStoreOption.setStoreLocatorId(newStoreLocator.getStoreLocatorId());
            newStoreOption.setName(entry.getKey());
            newStoreOption.setData(entry.getValue());
            storeOptionService.save(newStoreOption);
        }
        return new ErStore(inputStoreLocator, newPartitions, newOptions);
    }

    @Transactional
    public ErStore doDeleteStore(ErStore input) {
        ErStoreLocator inputStoreLocator = input.getStoreLocator();
        ErStoreLocator outputStoreLocator;
        if ("*".equals(inputStoreLocator.getName())) {
            UpdateWrapper<StoreLocator> updateWrapper = new UpdateWrapper<>();
            updateWrapper.lambda().setSql("name = concat(name, " + System.currentTimeMillis() + ")")
                    .set(StoreLocator::getStatus, Dict.DELETED)
                    .eq(StoreLocator::getNamespace, inputStoreLocator.getNamespace())
                    .eq(StoreLocator::getStatus, Dict.NORMAL)
                    .eq("*".equals(inputStoreLocator.getStoreType()), StoreLocator::getStoreType, inputStoreLocator.getStoreType());
            storeLocatorService.update(updateWrapper);
            outputStoreLocator = inputStoreLocator;
        } else {
            QueryWrapper<StoreLocator> queryWrapper = new QueryWrapper<>();
            queryWrapper.lambda().eq(StoreLocator::getStoreType, inputStoreLocator.getStoreType())
                    .eq(StoreLocator::getNamespace, inputStoreLocator.getNamespace())
                    .eq(StoreLocator::getName, inputStoreLocator.getName())
                    .eq(StoreLocator::getStatus, Dict.NORMAL);
            List<StoreLocator> nodeResult = storeLocatorService.list(queryWrapper);
            if (nodeResult.isEmpty()) {
                return null;
            }
            StoreLocator nodeRecord = nodeResult.get(0);
            String nameNow = nodeRecord.getName() + "." + System.currentTimeMillis();
            UpdateWrapper<StoreLocator> updateWrapper = new UpdateWrapper<>();
            updateWrapper.lambda().set(StoreLocator::getName, nameNow)
                    .set(StoreLocator::getStatus, Dict.DELETED)
                    .eq(StoreLocator::getStoreLocatorId, nodeRecord.getStoreLocatorId());
            storeLocatorService.update(updateWrapper);
            inputStoreLocator.setName(nameNow);
            outputStoreLocator = inputStoreLocator;
        }
        return new ErStore(outputStoreLocator, new ArrayList<>(), new ConcurrentHashMap<>());
    }

    public ErStoreList getStoreFromNamespace(ErStore input) {
        ErStoreList storeWithLocatorOnly = getStoreLocators(input);
        List<ErStore> newStores = new ArrayList<>();
        for (ErStore store : storeWithLocatorOnly.getStores()) {
            newStores.add(doGetStore(store));
        }
        storeWithLocatorOnly.setStores(newStores);
        return storeWithLocatorOnly;
    }

    public ErStoreList getStoreLocators(ErStore input) {
        QueryWrapper<StoreLocator> queryWrapper = new QueryWrapper<>();
        ErStoreLocator storeLocator = input.getStoreLocator();
        String storeName = storeLocator.getName();
        String storeNamespace = storeLocator.getNamespace();
        queryWrapper.apply(!StringUtils.isBlank(storeName), " and name like " + storeName.replace('*', '%'))
                .lambda().eq(!StringUtils.isBlank(storeNamespace), StoreLocator::getNamespace, storeNamespace);
        List<StoreLocator> storeList = storeLocatorService.list(queryWrapper);
        List<ErStore> erStoreArr = new ArrayList<>();
        for (int i = 0; i < storeList.size(); i++) {
            StoreLocator store = storeList.get(i);
            ErStoreLocator erStoreLocator = new ErStoreLocator(store.getStoreLocatorId()
                    , store.getStoreType()
                    , store.getNamespace()
                    , store.getName()
                    , store.getPath()
                    , store.getTotalPartitions()
                    , store.getPartitioner()
                    , store.getSerdes());
            erStoreArr.add(new ErStore(erStoreLocator, new ArrayList<>(), new ConcurrentHashMap<>()));
        }
        return new ErStoreList(erStoreArr, new ConcurrentHashMap<>());
    }

}
